### ISSUE TITLE
mod_push does not trigger if publish-options are included

### ISSUE THREAD

**Issue by VictorSFaria** [processone/ejabberd#3708]
Before creating a ticket, please consider if this should fit the discussion forum better:
https://github.com/processone/ejabberd/discussions

## Environment

- ejabberd version: 21.07
- Erlang version: `erl +V`
- OS: Linux Amazon IAM2
- Installed from: from source

## Configuration (only if needed): grep -Ev '^$|^\s*#' ejabberd.yml

```yaml
loglevel: 4
...
```

## Errors from error.log/crash.log

No errors

## Bug description

I´ve enabled the push service in my client with publish options.
My XML looks like this:
11:15:17 SENT (0): <iq id='SxA63-37' type='set'><enable xmlns='urn:xmpp:push:0' jid='p2.solucoesluz.com' node='pushnode1234'><x xmlns='jabber:x:data' type='form'><field var='FORM_TYPE'><value>http://jabber.org/protocol/pubsub#publish-options</value></field><field var='secret'><value>Some secret content</value></field></x></enable></iq>

I'm geetig the following error:
[debug] <0.1264.0>@ejabberd_router:do_route/1:384 Route:
#iq{
 id =
  <<"rr-1636417767040-11488676356814195610-oQgEgzTUwF9OvolwKa6TUc19mYA=-32938678">>,
 type = error,lang = <<>>,
 from =
  #jid{
   user = <<"victor.faria">>,server = <<"chat.solucoesluz.com">>,
   resource = <<>>,luser = <<"victor.faria">>,
   lserver = <<"chat.solucoesluz.com">>,lresource = <<>>},
 to =
  #jid{
   user = <<>>,server = <<"chat.solucoesluz.com">>,resource = <<>>,
   luser = <<>>,lserver = <<"chat.solucoesluz.com">>,lresource = <<>>},
 sub_els =
  [#pubsub{
    subscriptions = undefined,subscription = undefined,
    affiliations = undefined,
    publish =
     #ps_publish{
      node = <<"1D7nC7ChObh5">>,
      items =
       [#ps_item{
         xmlns = <<>>,id = <<>>,
         sub_els =
          [#push_notification{
            xdata =
             #xdata{
              type = submit,instructions = [],title = undefined,
              reported = undefined,items = [],
              fields =
               [#xdata_field{
                 label = <<>>,type = hidden,var = <<"FORM_TYPE">>,
                 required = false,desc = <<>>,
                 values = [<<"urn:xmpp:push:summary">>],
                 options = [],sub_els = []},
                #xdata_field{
                 label = <<"The body text of the last received message">>,
                 type = 'text-single',var = <<"last-message-body">>,
                 required = false,desc = <<>>,
                 values = [<<"New message">>],
                 options = [],sub_els = []}]}}],
         node = <<>>,publisher = <<>>}]},
    publish_options =
     #xdata{
      type = submit,instructions = [],title = undefined,reported = undefined,
      items = [],
      fields =
       [#xdata_field{
         label = <<>>,type = undefined,var = <<"FORM_TYPE">>,required = false,
         desc = <<>>,
         values = [<<"http://jabber.org/protocol/pubsub#publish-options">>],
         options = [],sub_els = []},
        #xdata_field{
         label = <<>>,type = undefined,var = <<"secret">>,required = false,
         desc = <<>>,
         values = [<<"SKv5D+74CWxDoJmO7tYy/r9A">>],
         options = [],sub_els = []}]},
    subscribe = undefined,unsubscribe = undefined,options = undefined,
    items = undefined,retract = undefined,create = undefined,
    configure = undefined,default = undefined,delete = undefined,
    purge = undefined,rsm = undefined},
   #stanza_error{
    type = wait,code = 500,by = undefined,reason = 'resource-constraint',
    text =
     [#text{
       lang = <<"en">>,
       data =
        <<"Unknown field 'secret' of type 'http://jabber.org/protocol/pubsub#publish-options'">>}],
    sub_els = []}],
 meta = #{}}

<0.490.0>@mod_push:notify/5:514 victor.faria@chat.solucoesluz.com rejected notification for victor.faria@chat.solucoesluz.com (1D7nC7ChObh5) temporarily: Unknown field 'secret' of type 'http://jabber.org/protocol/pubsub#publish-options' (resource-constraint)

This look like the same as: https://github.com/processone/ejabberd/issues/2048

**Comment by licaon-kter** [comment]
Which push server do you use?

**Comment by weiss** [comment]
> This look like the same as: #2048

Indeed, see [this comment][1] and the following ones.  Short story: It looks like you're trying to build a push server on top of `mod_pubsub`, which won't work.  Just pretend the notifications sent by `mod_push` would be normal messages (I have no idea why [XEP-0357][2] uses PubSub-like syntax).

I'm closing this issue, as it's not an ejabberd bug.

[1]: https://github.com/processone/ejabberd/issues/2048#issuecomment-336244980
[2]: https://xmpp.org/extensions/xep-0357.html

**Comment by VictorSFaria** [comment]
> Which push server do you use?

I'm using https://github.com/iNPUTmice/p2

**Comment by weiss** [comment]
The client told `mod_push` to send push notifications to `victor.faria@chat.solucoesluz.com` by setting the `jid` field of [the `<enable/>` request][1] to that address.  It should specify the server JID (i.e., the host name / subdomain) of your `p2` instance instead.

[1]: https://xmpp.org/extensions/xep-0357.html#enabling

**Comment by VictorSFaria** [comment]
> > This look like the same as: #2048
> 
> Indeed, see [this comment](https://github.com/processone/ejabberd/issues/2048#issuecomment-336244980) and the following ones. Short story: It looks like you're trying to build a push server on top of `mod_pubsub`, which won't work. Just pretend the notifications sent by `mod_push` would be normal messages (I have no idea why [XEP-0357](https://xmpp.org/extensions/xep-0357.html) uses PubSub-like syntax).
> 
> I'm closing this issue, as it's not an ejabberd bug.

'modpush append the options to the message, so `modpush` do this is OK?

In [this comment] https://github.com/processone/ejabberd/issues/2048#issuecomment-336305848 is said that mod_push propose is deal with conversations, that is the app server that I'm using.

**Comment by weiss** [comment]
> conversations, that is the app server that I'm using.

Yes, your case is different from #2048 then.  You probably just need to fix the push server JID as mentioned [above][1].

[1]: https://github.com/processone/ejabberd/issues/3708#issuecomment-964077309

**Comment by licaon-kter** [comment]
@VictorSFaria You forked Conversations and published it on Google Play? Link?

**Comment by VictorSFaria** [comment]
> @VictorSFaria You forked Conversations and published it on Google Play? Link?

No, I'm using p2 push server used by conversations, my company is developing its own mobile app.
